text/XXXX-config_revision.md

```diff
@@ -0,0 +1,268 @@
+```
+
+### Expanding Store to Filesystems
+
+Currently we hard-code qri filesystem (`qfs`) configuration details. qfs spins up a multiplexed-filesystem by default, composing multiple file-like persistence layers using prefix demuxing. If we expand "Store" to "Filesystem", we can make this fully configurable. The default configuration qri is using would look like this:
+
+```yaml
+Filesystems:
+- type: ipfs
+  config:
+    path: ~/.ipfs
+    api: true
+    pubsub: false
+- type: local
+- type: http
+```
+
+The first win: users could remove `local` or `http` to disable qri's access to the local filesystem or HTTP for the purpose of reading data. We can also expand for later filesystems & options. an `s3` implementation would be nice. this is also how we'd do `dat` integration.
+
+We could use a convention of the default write destination being the first listed filesystem, and present a big warning if users configure anything other than `ipfs` in the first position.
+
+It'd be great to delegate this entire section of store setup to the qfs package, aliasing up the conguration data structure into the config package.
+
+### move IPFS repo location into configuration
+the above filesystem example shows a `ipfs` filesystem type with a `config.path` field. This value should now the be canonical source of where to load.
+
+the global `ipfs-path` flag should also be removed:
+```
+      --ipfs-path string   override IPFS path location (default "/Users/b5/.ipfs")
+```
+
+### Relative Paths in configuration files are relative to the file
+If a configuration value specifies a filepath, and the given path is relative, that path should be relative to the location of the config file.
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+### Per-field env vars
+We could support setting configuration on a per-field basis. I'm not sure that's totally necessary at this point.
+
+### Attach config to context.Context
+We could build a sophisticated system of attaching configuration to request contexts, to support scoping configuration on a per-request basis. We might need to in the future, but it's overkill for the moment.
+
+### Config changes emit events
+We could also use the event bus to propagate changes to configuration. Again, overkill for now, but we could do so later.
+
+# Prior art
+[prior-art]: #prior-art
+
+### Configuration:
+awesome-go has an entire section of configuration libraries https://awesome-go.com/#configuration. Some of the ones I found intersting:
+* hot-reloading config: https://github.com/lalamove/konfig
+* env var config with lots of bells& whistles: https://github.com/kelseyhightower/envconfig
+
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+Some configuration things are out of scope for this RFC, but should be discussed later:
+* Distinguish runtime _options_ from static _configuration_.
+* hot-swapping configuration.
```

Comment:
Think all operations relying on config should read from config directly for each time they run. Thing's that rely on config on startup should not be hot-swappable IMHO and should require an instance restart. Thus hot-swapping should be a given if we properly utilize the new config store.

Reply:
If desktop is running, this means that any time user switches a flag from local to network or (when we create a settings page) edits any settings that affect the backend, we would have to kill the qri process and then reconnect, or the user would have shut down the app and restart. Which... might be fine, especially the former option and when talking about general settings, but honestly toggling between `local` and `network` feels like something user will do a lot, casually, and we should make it seamless for them.

Reply:
If desktop is running, this means that any time user switches a flag from local to network or (when we create a settings page) edits any settings that affect the backend, we would have to kill the qri process and then reconnect, or the user would have shut down the app and restart. Which... might be fine, especially the former option and when talking about general settings, but honestly toggling between local and network feels like something user will do a lot, casually, and we should make it seamless for them.

Reply:
I think the first time we ship an on-off switch for p2p on desktop, it should be a fairly dramatic action that locks you out of desktop during the restart. This is one part UX, one part being conservative with network churn.

We want to keep people leaving and joining p2p networks lower if we can. Making it too easy to swap p2p will lead to people joining and leaving the network more than they need to. The added upside is an easier story around configuration. If anyone _really_ complains about p2p switching, we can re-evaluate.